jobs/search-term-data-validation/src/data_validation.py

```diff
@@ -0,0 +1,421 @@
+
+    Inputs:
+
+    - validation_data: the dataframe with the data in it to be checked.
+    ASSUMES the presence of a 'finished_at' column, whose date is used to calculate lookback and test windows.
+    - metric: the name of the column in the input dataframe on which to perform the check.
+    - full_lookback_window: an integer number of days that the comparison set should cover.
+    - test_window. an integer number of days that the test set should cover.
+    ASSUMES that the test window immediately succeeds the full_lookback_window.
+    - moving_average_window: an integer. Number of prior days over which to calculate an average for a given day.
+    - mean lower bound: a float between 0 and 1 indicating the lower edge of the window of normal values from the comparison set
+    inside which at least one of the values in the test set should fall.
+    - mean upper bound: a float between 0 and 1 indicating the upper edge of the window of normal values from the comparison set
+    inside which at least one of the values in the test set should fall.
+
+
+    Outputs:
+    - finished_at: the finished_at timestamp of the job run to which this check applies.
+    - num_moving_averages_compared: an integer representing the total number of moving average values included in this comparison.
+    - should_trigger: a bool indicating whether the values in the test window are all falling OUTSIDE the expected range.
+    - mean_lower: a float. The lower bound of the expected range of moving averages calculated from comparison values.
+    - mean_upper: a float. The upper bound of the expected range of moving averages calculated from comparison values.
+    - moving_average_windo: an integer. The moving average window passed into the function.
+    - test_moving_averages: a list. The entirety of the test values.
+
+    """
+    if not (0 < mean_lower_bound < 1 and 0 < mean_upper_bound < 1):
+        raise Exception(
+            "mean_lower_bound and mean_upper_bound should both be between zero (0) and one (1)."
+        )
+
+    if "finished_at" not in validation_data.columns.values:
+        raise Exception("dataframe must include a finished_at column.")
+
+    if metric not in validation_data.columns.values:
+        raise Exception(f'dataframe does not include target metric "{metric}"')
+
+    today = date.today()
+    latest_finished_at = max(validation_data["finished_at"])
+
+    test_earliest_date = today - timedelta(days=test_window)
+    comparison_earliest_date = test_earliest_date - timedelta(days=full_lookback_window)
+
+    x_day_moving_average = f"{moving_average_window}_day_{metric}_moving_avg"
+    validation_data[x_day_moving_average] = (
+        validation_data[metric]
+        .rolling(window=moving_average_window, min_periods=0)
+        .mean()
+    )
+
+    comparison_values = validation_data["finished_at"].apply(
+        lambda m: comparison_earliest_date < m.date() <= test_earliest_date
```

Comment:
I'm not sure what your dataset size is but maybe worth considering np.where? Definitely not a big enough deal to prevent approval from me though